File: leetcode_py/House Robber III.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

def depthFirst(root):
    if root == None:
        return 0

    root.val += depthFirst(root.left)
    root.val += depthFirst(root.right)
     
    next = (root.left.val if root.left!=None else 0) +  (root.right.val if root.right else 0)
    root.val = max(next, root.val);
    return next;


class Solution(object):
    def rob(self, root):
        if not root:
            return 0
        """
        :type root: TreeNode
        :rtype: int
        """
        depthFirst(root)
        return root.val

if __name__=="__main__":
    root = TreeNode(3)
    root.left = TreeNode(4)
    root.right = TreeNode(5)
    root.left.left = TreeNode(1)
    root.left.right = TreeNode(3)
    root.right.right = TreeNode(1)
    a = Solution()
    print a.rob(root)

File: leetcode_js/House Robber III.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */


function depthFirst(root){
    if(root == undefined) return 0;

    root.val += depthFirst(root.left);
    root.val += depthFirst(root.right);
     
    next = (root.left!=undefined?root.left.val:0)+(root.right!=undefined?root.right.val:0)
    root.val = Math.max(next, root.val);
    return next;
}
var rob = function(root) { 
    if(root == undefined) return 0;
    depthFirst(root);
    return root.val;
};


 // this will TLE
 
function depthFirst(root){
    if(root == undefined) return 0;
    var res_y = root.val;
    if(root.left != undefined) {
        res_y += depthFirst(root.left.left);
        res_y += depthFirst(root.left.right);
    }
    if(root.right != undefined){
       res_y += depthFirst(root.right.left);
       res_y += depthFirst(root.right.right);
    }
   
    var res_n = 0;
    res_n += depthFirst(root.left);
    res_n += depthFirst(root.right);
    return Math.max(res_y, res_n);
}
  

 
